### PR TITLE
Add plot related tests

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Add more tests to cover corner case scenarios (mostly plotting related).

# Details

This is pulled out of #1516 and adds coverage for corner cases. Note that some of these are not needed with the main branch but could be relevant if we change things in the future (such as #1516), but some areas are definitely new (e.g. they show some issues discussed in #1530).

As we have a number of plot-related changes coming in, it is useful to have these corner cases covered sooner rather than later.

### CODE REVIEW

File: sherpa/astro/ui/tests/test_astro_session.py [hunk 1]
```diff
@@ -406,3 +413,815 @@ def test_save_error_ascii(session, kwargs, idval, tmp_path):
...
+    #
+    s.set_method("levmar")
+
+    out = StringIO()
+    s.show_method(outfile=out)
+
+    toks = out.getvalue().split("\n")
+    assert toks[0] == "Optimization Method: LevMar"
+    assert toks[1] == "name     = levmar"
+    assert toks[2] == "ftol     = 1.1920928955078125e-07"
```
Comment (hamogu) [inline sherpa/astro/ui/tests/test_astro_session.py:497]: I'm glad this works, but I'm always wary of hardcoding so much precision. It's prone to failing and creating confusion (e.g. if some user runs it on some other hardware or python version) or on totally unrelated changes because some default numpy printing changed. I suggest to check only what is the minimum we expect for this to make sense, e.g. `assert re.match("1.19[0-9]*e-07")`. We want to test to fail is this is much more or less than 1e-7 or of the formatting of the print statement changes significantly, but we don't need it to fail because the 17th digit is either not printed to differs in some way.

While we have some tests that are classic regression tests ("Make sure we manually check every change"), here I would advocate for a setup that's a little more lenient. If this was not a string comparison, we would use `pytest.approx` for the same effect.

File: sherpa/astro/ui/tests/test_astro_session.py [hunk 2]
```diff
@@ -406,3 +413,815 @@ def test_save_error_ascii(session, kwargs, idval, tmp_path):
...
+
+    # The following is somethiing that will need to be changed
+    # whenever an underlying object has it's string representation
+    # changed, but it should not be hard to do. The main issue is
+    # going to be numeric precision.
+    #
+    toks = out.getvalue().split("\n")
+    assert toks[0] == "Optimization Method: NelderMead"
+    assert toks[1] == "name         = simplex"
+    assert toks[2] == "ftol         = 1.1920928955078125e-07"
```
Comment (hamogu) [inline sherpa/astro/ui/tests/test_astro_session.py:574]: Same as comment above.

File: sherpa/astro/ui/tests/test_astro_session.py [hunk 3]
```diff
@@ -406,3 +413,815 @@ def test_save_error_ascii(session, kwargs, idval, tmp_path):
...
+
+    mclass = getattr(sherpa.plot, f"{label.capitalize()}Plot")
+    assert isinstance(mplot, mclass)
+
+    cclass = getattr(sherpa.plot, f"Component{label.capitalize()}Plot")
+    assert isinstance(cplot, cclass)
+
+    # Check it's not a sub-class.
+    #
+    subclass = getattr(sherpa.plot, f"ComponentTemplate{label.capitalize()}Plot")
```
Comment (hamogu) [inline sherpa/astro/ui/tests/test_astro_session.py:855]: This is an example where it's not clear to me why this is worth testing. Why could this not be a sub-class? (No need to explain this to me, just read on and click "resolve conversation".) However, that's fine for a Regression test. It passes now, so it clearly tests existing behavior. If it ever changes, someone will have to confirm that ti's an intended change.

**Review by hamogu — CHANGES_REQUESTED** [review]
This is basically ready to approve. I admit that I only skimmed over some of the changes and did not check every line. However, I think in this case, that's sufficient. The vast majority of this PR is some simply "Regression test - the intended behavior is unclear". As such, any test that passes now works - since it works, it's obviously checking current behavior. 

My comments highlight an example where I decided that the fact that it tests something and passes now is good enough without me digging deeper *why* this or that happens. I also have one suggestion to make it more robust to running tests on different architectures.

File: sherpa/astro/ui/tests/test_astro_session.py (hunk 3 above)
Comment (DougBurke) [reply sherpa/astro/ui/tests/test_astro_session.py:855]: It is a regression test. I don't understand the template plotting code and what it should do, so I'm just testing what it does do now and leaving it to #1530 for future discussion.

File: sherpa/astro/ui/tests/test_astro_session.py (hunk 1 above)
Comment (DougBurke) [reply sherpa/astro/ui/tests/test_astro_session.py:497]: We have quite a few tests that I expect to fail at some point if there's significant numerical differences. I've actually been quite surprised we haven't seen many already. I think that it's in part many of these values (like the 1.19..e-7) value are "special" - e.g.

```
>>> numpy.finfo(numpy.float32).eps
1.1920929e-07
```

and so they don't suffer from "'rounding/..." issues that a value that was calculated would have. There's still the issue of how-many decimal places get displayed, but I think that is either a python or numpy decision (depending on exactly how it's created).

So in this case I don't really have the energy to change these lines until we start seeing a problem with them, but there are other places where we could be more defensive.

**Review by hamogu — APPROVED** [review]

**Review by hamogu — APPROVED** [review]
Still fine with me.
